sys/net/link_layer/ble_ll/ble_ll.c

```diff
@@ -0,0 +1,334 @@
+{
+    /* check parameters */
+    if (len != BLE_ADV_ADDR_LEN) {
+        return -EINVAL;
+    }
+
+    /* set address */
+    _adv_addr[0] = val[5];
+    _adv_addr[1] = val[4];
+    _adv_addr[2] = val[3];
+    _adv_addr[3] = val[2];
+    _adv_addr[4] = val[1];
+    _adv_addr[5] = val[0];
+
+    return BLE_ADV_ADDR_LEN;
+}
+
+/* Set netopt parameter value in current BLE interface configuration */
+static int _set(gnrc_netdev_t *dev, netopt_t opt,
+         void *value, size_t value_len)
+{
+    switch (opt) {
+        case NETOPT_BLE_ADV_ADDRESS:
+            return _set_adv_address(value, value_len);
+
+        default:
+            return dev->driver->set(dev, opt, value, value_len);
+    }
+}
+
+/* Get netopt parameter value from current BLE interface configuration */
+static int _get(gnrc_netdev_t *dev, netopt_t opt,
+         void *value, size_t max_len)
+{
+    switch (opt) {
+        case NETOPT_BLE_ADV_ADDRESS:
+            return _get_adv_address(value, max_len);
+
+        default:
+            return dev->driver->get(dev, opt, value, max_len);
+    }
+}
+
+/**
+ * @brief   BLE link layer send function
+ */
+static int _send(gnrc_netdev_t *dev, gnrc_pktsnip_t *pkt)
+{
+    uint8_t payload[BLE_PDU_LEN_MAX];
+    gnrc_pktsnip_t *ble_pkt = NULL;
+    uint16_t pkt_len;
+    uint8_t payload_len;
+
+    if (pkt == NULL) {
+        DEBUG("ble_ll: Error handling packet: packet incomplete\n");
+        return -ENOMSG;
+    }
+
+    pkt_len = gnrc_pkt_len(pkt);
+
+    if (pkt_len > BLE_PDU_LEN_MAX) {
+        gnrc_pktbuf_release(pkt);
+        DEBUG("ble_ll: Error sending packet: invalid BLE pdu header length\n");
+        return -EOVERFLOW;
+    }
+
+    /* write advertising pdu header to payload */
+    memcpy(payload, pkt->data, BLE_ADV_HDR_LEN);
+    /* write advertising address to payload */
+    memcpy((payload + BLE_ADV_HDR_LEN), _adv_addr, BLE_ADV_ADDR_LEN);
+    pkt_len += BLE_ADV_ADDR_LEN;
+
+    /* write optional payload data */
+    if (pkt->next != NULL) {
+        payload_len = pkt->next->size;
+        if (payload_len > BLE_ADV_PAYLOAD_LEN_MAX) {
+            gnrc_pktbuf_release(pkt);
+            DEBUG("ble_ll: Error handling packet: advertising payload too large\n");
+            return -EOVERFLOW;
+        }
+        memcpy((payload + BLE_ADV_HDR_LEN + BLE_ADV_ADDR_LEN),
+                pkt->next->data, payload_len);
+    }
+
+    ble_pkt = gnrc_pktbuf_add(NULL, payload, pkt_len, GNRC_NETTYPE_UNDEF);
+    gnrc_pktbuf_release(pkt);
+
+    return dev->driver->send_data(dev, ble_pkt);
+}
+
+/**
+ * @brief    Handle incoming link layer packet
+ */
+static void _receive(gnrc_pktsnip_t *pkt)
+{
+    ble_adv_pkt_t *data;
+
+    data = (ble_adv_pkt_t *)pkt->data;
+
+    if (_ll_state == LL_ADVERTISING) {
+        /* Send respone to received BLE SCAN_REQ packet */
+        if (data->header.pdu_type == SCAN_REQ_TYPE) {
+            DEBUG("ble_ll: receive_data -> SCAN_REQ received\n");
+            /*_send_scanresponse_pkt();*/
```

Comment:
I suppose. I left it in to show what will happen in this function later (since it doesn't really do anything right now). Think I should delete it for now?